tensorflow/core/kernels/stack.cc

```diff
@@ -72,19 +80,96 @@ class Stack : public ResourceBase {
       return errors::InvalidArgument("Stack[", stack_name_,
                                      "] is empty when calling Pop().");
     }
-    *value = stack_.back();
+    int index = stack_.size() - 1;
+    while (is_swapping_ins_[index]) {
+      cond_swapping_ins_[index].wait(l);
+      index = stack_.size() - 1;
+    }
+    *value = stack_.back().first;
+    auto iter = stack_.back().second;
+    for (auto i = unswapped_lru_.begin(); i != unswapped_lru_.end(); i++) {
+      if (i == iter) {
+        unswapped_lru_.erase(iter);
+        break;
+      }
+    }
     stack_.pop_back();
     return Status::OK();
   }
 
+  Status MarkAsSwapped(TensorAndAllocation* value, Tensor* cpu_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *cpu_tensor;
+    value->swapped_to_cpu = true;
+
+    return Status::OK();
+  }
+
+  Status MarkAsUnswapped(TensorAndAllocation* value, Tensor* device_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *device_tensor;
+    value->swapped_to_cpu = false;
+
+    return Status::OK();
+  }
+
+  bool GetTensorToSwapOut(
+      TensorAndAllocation** value,
+      std::function<bool(const TensorAndAllocation&)> cond) {
+    mutex_lock l(mu_);
+    bool can_swap = false;
+    while (!can_swap) {
+      if (unswapped_lru_.empty()) {
+        break;
+      }
+      int index = unswapped_lru_.front();
+      unswapped_lru_.pop_front();
```

Comment:
Same as before: whenever we adjust something, can we keep alll the data structures consistent with one another? In this case, if an element is removed from unswapped_lru_, we should adjust the iterator in stack_ to be unset.

You can use absl::optional for the iterator for that purpose (it should have the same API as the new std::optional in C++17).

Reply:
Thanks for the neat idea. I've pushed the latest commit adhering this, is it in line with what you had in mind?

Reply:
Yeah, but if we're going to keep the swapping from CPU to GPU, we'll need to keep swapped_mru_ consistent as well, meaning stack_ should be a std::tuple of 3 things:
1. The TensorAndAllocation.
2. optional iterator into unswapped_lru_
3. optional iterator into swapped_mru_

Popping an element should also try to erase from swapped_mru_. Basically we should never need to check if the index in unswapped_lru_ or swapped_mru_ is valid, because we guarantee that all of stack_, unswapped_lru_, and swapped_mru_ are consistent with one another.

Reply:
> Thanks for your patience. It's looking much better. Just some revisions.

Thanks! I'm learning so much from these reviews alone.

Alright, in the latest commit I've kept `swapped_mru_` consistent as well. Now the index checks are gone, and they're working well.
Anyway, I put them in `StackElement` struct instead of std::tuple, since the type is getting really complicated. But if you prefer, I'd change them to std::tuple.